decidim-core/app/services/decidim/activity_search.rb

```diff
@@ -120,6 +121,50 @@ def filter_follows(query)
       query.where(chained_conditions)
     end
 
+    def filter_hidden(query)
+      # Filter out the items that have been moderated.
+      query = query.joins(
+        <<~SQL.squish
+          LEFT JOIN decidim_moderations
+            ON decidim_moderations.decidim_reportable_type = decidim_action_logs.resource_type
+            AND decidim_moderations.decidim_reportable_id = decidim_action_logs.resource_id
+            AND decidim_moderations.hidden_at IS NOT NULL
+        SQL
+      ).where(decidim_moderations: { id: nil })
+
+      # Filter out the private space items that should not be visible for the
+      # current user.
+      current_user_id = options.fetch(:current_user, nil)&.id.to_i
+      Decidim.participatory_space_manifests.map do |manifest|
+        klass = manifest.model_class_name.constantize
+        next unless klass.include?(Decidim::HasPrivateUsers)
+
+        table = klass.table_name
+        query = query.joins(
+          Arel.sql(
+            <<~SQL.squish
+              LEFT JOIN #{table}
+                ON decidim_action_logs.participatory_space_type = '#{manifest.model_class_name}'
+                AND #{table}.id = decidim_action_logs.participatory_space_id
+              LEFT JOIN decidim_participatory_space_private_users AS #{manifest.name}_private_users
+                ON #{manifest.name}_private_users.privatable_to_type = '#{manifest.model_class_name}'
+                AND #{table}.id = #{manifest.name}_private_users.privatable_to_id
+            SQL
+          )
+        ).where(
+          Arel.sql(
+            <<~SQL.squish
+              #{table}.id IS NULL OR
+              #{table}.private_space = 'f' OR
+              #{manifest.name}_private_users.decidim_user_id = #{current_user_id}
+            SQL
+          )
+        )
+      end
+
+      query
```

Comment:
💬 I'm a bit worried about the performance and the mantainability of these queries, but I can't find a better approach, so let's merge it and see how it works. :man_shrugging: 😕

Reply:
@leio10

### Regarding performance

Generally performance is not an issue to worry about unless the query takes a really long time to execute and it is run very often. This query falls into neither of those categories.

This change will obviously make the query less performant than the original but here we are only talking about some milliseconds which is nothing.

I ran `EXPLAIN ANALYZE` on both of these queries on an instance with a short of 90k users and a user that has over 3k action log items in total or 1.5k public action log items that would be visible there.

The results are:

| Query | Planning time | Execution time | Total time |
|----------|--------------------|---------------------|---------------|
| Original | 0.355ms | 0.064ms | 0.419ms |
| Modified (PR) | 5.560ms | 0.230ms | 5.790ms |
| Difference | 5.205ms (or 15.66x slower) | 0.166ms (or 3.59x slower) | 5.371ms (or 13.82x slower) |

So can you see it's significantly slower (~14-15 times slower) but we are only talking about a difference of ~5ms here. It's not going to be noticeable.

---

### Regarding maintainability

You are correct that it makes it more difficult to maintain but I don't know another way which wouldn't be significantly less performant than the proposed solution. Such as pre-counting the visibility for all records which would take potentially a very long time with a lot of records. Maybe the private space activities should be completely hidden from the activity log (?), which would make it easier, but then it would change existing behavior.

For moderated action log items you could create some logic that would change a flag in the action logs table when the item is hidden by moderation. But I think that's conceptually wrong because then the action log items would have to be touched afterwards (I think the idea is to keep them intact).

For the private space items, I cannot think of a better way unless we create some kind of action log - user mapping indicating the visibility of the item per user. But that seems cumbersome and quite a lot more work as well.